datastore/providers/opensearch_datastore.py

```diff
@@ -0,0 +1,422 @@
+import os
+import json
+
+
+import boto3
+from opensearchpy import AsyncOpenSearch, AWSV4SignerAsyncAuth, AsyncHttpConnection
+from requests_kerberos import HTTPKerberosAuth, OPTIONAL
+
+from typing import Any, Dict, List, Optional
+from datastore.datastore import DataStore
+from models.models import (
+    DocumentChunk,
+    DocumentChunkWithScore,
+    DocumentMetadataFilter,
+    QueryResult,
+    QueryWithEmbedding, DocumentChunkMetadata,
+    Source
+)
+from services.date import validate_date_str
+
+# Read environment variables for OpenSearch
+OPENSEARCH_HOST = os.environ.get("OPENSEARCH_HOST", "localhost")
+OPENSEARCH_PORT = int(os.environ.get("OPENSEARCH_PORT", 9200))
+OPENSEARCH_USERNAME = os.environ.get("OPENSEARCH_USERNAME")
+OPENSEARCH_PASSWORD = os.environ.get("OPENSEARCH_PASSWORD")
+OPENSEARCH_AUTH_TYPE = os.environ.get("OPENSEARCH_AUTH_TYPE", "no-auth")
+OPENSEARCH_INDEX_NAME = os.environ.get("OPENSEARCH_INDEX_NAME", "open-ai-index")
+OPENSEARCH_KNN_FIELD_NAME = os.environ.get("OPENSEARCH_KNN_FIELD_NAME", "embedding")
+OPENSEARCH_INDEX_PRIMARIES = int(os.environ.get("OPENSEARCH_INDEX_PRIMARIES", 1))
+OPENSEARCH_INDEX_REPLICAS = int(os.environ.get("OPENSEARCH_INDEX_REPLICAS", 0))
+OPENSEARCH_KNN_ENGINE = os.environ.get("OPENSEARCH_KNN_ENGINE", "nmslib")
+OPENSEARCH_KNN_VECTOR_DISTANCE = os.environ.get("OPENSEARCH_KNN_VECTOR_DISTANCE", "l2")
+OPENSEARCH_EMBEDDING_IN_RESULT = bool(os.environ.get("OPENSEARCH_EMBEDDING_IN_RESULT", False))
+CA_CERTS_FULL_PATH = os.environ.get("CA_CERTS_FULL_PATH")
+AWS_REGION = os.environ.get("AWS_REGION", "us-east-2")
+CLOUD_SEARCH_SERVICE = os.environ.get("CLOUD_SEARCH_SERVICE", "es")  # 'aoss' for OpenSearch Serverless
+
+assert OPENSEARCH_HOST is not None
+assert OPENSEARCH_PORT is not None
+
+# OpenAI Ada Embeddings Dimension
+VECTOR_DIMENSION = 1536
+# This is a good enough number of size.
+OPENSEARCH_BULK_BATCH_SIZE = 200
+
+index_body = {
+    "settings": {
+        "index.knn": "true",
+        "number_of_shards": OPENSEARCH_INDEX_PRIMARIES,
+        "number_of_replicas": OPENSEARCH_INDEX_REPLICAS
+    },
+    "mappings": {
+        "properties": {
+            OPENSEARCH_KNN_FIELD_NAME: {
+                "type": "knn_vector",
+                "dimension": VECTOR_DIMENSION,
+                "method": {
+                    "name": "hnsw",
+                    "engine": OPENSEARCH_KNN_ENGINE
```

Comment:
Do we want to make parameters configurable?

Reply:
OPENSEARCH_KNN_ENGINE : is there for configuration. I can attach more configuration, but seeing that this is there to create a default index, I skipped providing multiple configuration. 

If customer needs more configuration he can create an index by himself.